mkdocs/config/config_options.py

```diff
@@ -0,0 +1,333 @@
+import os
+
+import six
+
+from mkdocs import utils, legacy
+from mkdocs.legacy import OrderedDict
+
+
+class ValidationError(Exception):
+    """Raised during the validation process of the config on errors."""
+
+
+class BaseConfigOption(object):
+    """
+    The BaseConfigOption adds support for default values and required values
+
+    It then delegates the validation and (optional) post processing to
+    subclasses.
+    """
+
+    def __init__(self, default=None, required=False):
+        self.default = default
+        self.required = required
+        self.warnings = []
+
+    def is_required(self):
+        return self.required
+
+    def validate(self, value):
+        """
+        Perform some initial validation.
+
+        If the option is empty (None) and isn't required, leave it as such. If
+        it is empty but has a default, use that. Finally, call the _validate
+        method on the subclass unless.
+        """
+
+        if value is None:
+            if self.default is not None:
+                value = self.default
+            elif not self.required:
+                return
+            elif self.required:
+                raise ValidationError("Required configuration not provided.")
+
+        return self._validate(value)
+
+    def _validate(self, value):
+        """
+        Perform validation for a value.
+
+        The _validate method should be implemented by subclasses.
+        """
+        return value
+
+    def post_process(self, config, key_name):
+        """
+        After all options have passed validation, perform a post process to
+        do any additional changes dependant on other config values.
+
+        The post process method should be implemented by subclasses.
+        """
+
+
+class Type(BaseConfigOption):
+    """
+    Type Config Option
+
+    Validate the type of a config option against a given Python type.
+    """
+
+    def __init__(self, type_, length=None, **kwargs):
+        super(Type, self).__init__(**kwargs)
+        self._type = type_
+        self.length = length
+
+    def _validate(self, value):
+
+        if not isinstance(value, self._type):
+            msg = ("Expected type: {0} but recieved: {1}"
+                   .format(self._type, type(value)))
+        elif self.length is not None and len(value) != self.length:
+            msg = ("Expected type: {0} with lenght {2} but recieved: {1} with "
+                   "length {3}").format(self._type, value, self.length,
+                                        len(value))
+        else:
+            return value
+
+        raise ValidationError(msg)
+
+
+class URL(BaseConfigOption):
+    """
+    URL Config Option
+
+    Validate a URL by requiring a scheme is present.
+    """
+
+    def _validate(self, value):
+
+        parsed_url = six.moves.urllib.parse.urlparse(value)
```

Comment:
Should this be in a try/except block? If `urlparse` fails, what error do you want to be given?